### ISSUE TITLE
Encryption of Opus audio failing

### ISSUE THREAD

**Issue by dominic-p** [gpac/gpac#1476]
Thanks for reporting your issue. Please make sure these boxes are checked before submitting your issue - thank you!

- [x] I looked for a similar issue and couldn't find any.
- [x] I tried with the latest version of GPAC. Installers available at http://gpac.io/downloads/gpac-nightly-builds/
- [x] I give enough information for contributors to reproduce my issue (meaningful title, github labels, platform and compiler, command-line ...). I can share files anonymously with this dropbox: https://www.mediafire.com/filedrop/filedrop_hosted.php?drop=eec9e058a9486fe4e99c33021481d9e1826ca9dbc242a6cfaab0fe95da5e5d95

Detailed guidelines: http://gpac.io/2013/07/16/how-to-file-a-bug-properly/

I'm trying to get ClearKey encryption of a DASH stream working. I have the video working well, but when it comes to the audio files, I'm running into an error.

Here are the steps to reporduce.

1. Get an Opus audio file (this [example one](https://people.xiph.org/~giles/2012/opus/ehren-paper_lights-96.opus) from the opus website works)

2. Add the Opus audio to an MP4 container

```
MP4Box -add ehren-paper_lights-96.opus opus.mp4
```

3. Attempt to encrypt the file

```
MP4Box -crypt drm_file.xml opus.mp4 -out opus-enc.mp4
```

Using this drm_file.xml
```
<GPACDRM type="CENC AES-CTR">
  <DRMInfo type="pssh" version="1">
    <BS ID128="1077efecc0b24d02ace33c1e52e2fb4b"/>
    <BS bits="32" value="1"/>
    <BS ID128="cd7eb9ff88f34caeb06185b00024e4c2"/>
  </DRMInfo>
  <CrypTrack IV_size="8" first_IV="0xbb5738fe08f11341" isEncrypted="1" saiSavedBox="senc" trackID="1">
    <key KID="0xcd7eb9ff88f34caeb06185b00024e4c2" value="0x63cb5f7184dd4b689a5c5ff11ee6a328"/>
  </CrypTrack>
</GPACDRM>
```

This yields an error:
```
[CENC] Cannot create CENC protection: Bad Parameter

        Error: Bad Parameter
```

Is encrypted Opus audio not supported? Or, is there a work around that I don't know about?

I tried adding the video and audio to the same file and then encrypting that. But, when I dashed it the MPD only had a `<ContentProtection ... />` element in the video `<AdapationSet ... />` and not the audio one, so I'm guessing the audio was never encrypted.

**Comment by dominic-p** [comment]
As a side note I tried building the latest code from the filters branch and got pretty much the same error:

```
[MP4Mux] Failed to setup CENC information: Bad Parameter
Segmentation fault (core dumped)
```

```
$ MP4Box -version
MP4Box - GPAC version 0.9.0-DEV-rev91-ge6b9d9811-filters
(c) 2000-2020 Telecom Paris distributed under LGPL v2.1+ - http://gpac.io
GPAC Configuration: --static-mp4box
Features: GPAC_CONFIG_LINUX GPAC_64_BITS GPAC_HAS_IPV6 GPAC_HAS_SOCK_UN GPAC_MINIMAL_ODF GPAC_HAS_QJS GPAC_HAS_LINUX_DVB  GPAC_DISABLE_3D 
```

**Comment by jeanlf** [comment]
Indeed we had issues with opus, flac and other. Fixed on git, code should now be more generic for CENC setup, thanks for the report.

**Comment by dominic-p** [comment]
Thanks so much! That did it. :)

**Comment by dominic-p** [comment]
Actually, I spoke too soon. :( I can encrypt the opus.mp4 now, but when I try to DASH it, I'm getting a segmentation fault. I also see a bunch of errors if I try to add the .opus file to an MP4 container with the 0.9 version (the same command with the 0.8 branch completes without errors):

```
$ MP4Box -add ehren-paper_lights-96.opus opus.mp4
...
[MPEG-2 TS] TS Packet 216 does not start with sync marker
[MPEG-2 TS] TS Packet 217 does not start with sync marker
[MPEG-2 TS] TS Packet 218 is scrambled - not supported
[MPEG-2 TS] TS Packet 219 does not start with sync marker
[MPEG-2 TS] TS Packet 220 does not start with sync marker
[MPEG-2 TS] TS Packet 221 has error (PID could be 2432)
[MPEG-2 TS] TS Packet 222 does not start with sync marker
[MPEG-2 TS] TS Packet 223 does not start with sync marker
...
```

I'm also running into similar issues (segmentation faults, bad parameter, etc.) with AV1 video. It's strange, I was able to DASH AV1 video and Opus audio no problem using the 0.8 branch. But, once I tried to add DRM into the mix everything started failing.

Maybe I need to wait a bit for the 0.9 branch to get closer to release before trying to work with encrypted AV1 and Opus? I can open another issue if it would be helpful. Just let me know.

**Comment by jeanlf** [comment]
Indeed we had a bug on handling opus in mp4 with implications on dasher, should now be fixed. Could you check ?

I cannot reproduce your import bug, are you sure you tried this one with 0.9 ?

As a side note, we're very close to releasing so any bug you found on 0.9 is worth reporting !

**Comment by dominic-p** [comment]
Thanks for the follow up. I'm no longer getting errors dashing the file. However, I'm having trouble testing it because it seems like the behavior of the `-segment-name` flag has changed.

[This wiki entry](https://github.com/gpac/gpac/wiki/DASH-Introduction#multiple-representations) suggests using the `%s` template to prevent segments from overwriting each other. However, when I do that in 0.9 a literal `%s` is put in the file name and the segments are overwritten. Is that a bug or has that feature been removed?

**Comment by jeanlf** [comment]
That was a bug, we forgot to support old syntax. This should now be fixed, thanks for the report.

**Comment by dominic-p** [comment]
Thanks! I can confirm that syntax is working now. Now, I'm getting the following errors when I attempt to dash some AV1 video with the Opus audio:

```
[AV1] incoherent color characteristics primaries 2 transfer 2 matrix 2 color range 0
[Dasher] Segments do not start with the same SAP types: set initialized with 1 but first packet got 4 - bitstream will not be compliant
```

The incoherent color seems like more of a warning, but the SAP types seems like problem. I'm guessing it's a problem with how I encoded the video or audio, but I wasn't able to find an explanation of what that means. Is there documentation on this I can read?

**Comment by jeanlf** [comment]
I think there's a audio roll somewhere leading to SAP4 packets, but this should be ignored by default with dasher.strict_sap=off option.
Could you upload your sample files somewhere ?

**Comment by dominic-p** [comment]
Thanks for taking a look. I uploaded the files to the mediafire link in the issue template. Here is the exact command I'm using:

```
MP4Box -dash 3000 -frag 3000 -rap -bs-switching no -segment-name seg_%s -segment-timeline -url-template -out av1/video-av1.mpd video-av1-1920.mp4 video-av1-1280.mp4 video-av1-768.mp4 video-av1-512.mp4 audio-opus.mp4 audio-opus-low.mp4
```

**Comment by jeanlf** [comment]
Both AV1 color report and SAP mismatch should now be fixed.

Regarding your shaka player issue, have you tried without protection ?

**Comment by dominic-p** [comment]
Thanks for the update. I have confirmed that the DASH command above completes without errors now.

As for the shaka player issue, I have tried it with the same files and the same version of MP4Box only skipping the encryption step and the stream plays no problem.

I'm setting up the DRM like this in shaka (see [tutorial here](https://shaka-player-demo.appspot.com/docs/api/tutorial-drm-config.html)).

```
player.configure({
  drm: {
    clearKeys: {
      // 'key-id-in-hex': 'key-in-hex',
      'deadbeefdeadbeefdeadbeefdeadbeef': '18675309186753091867530918675309'
    }
  }
});
```

The only error or warning I see in the console is:

> AbortError: The fetching process for the media resource was aborted by the user agent at the user's request.
